src/webgpu/util/texture/layout.ts

```diff
@@ -0,0 +1,194 @@
+import * as C from '../../../common/constants.js';
+import { assert, unreachable } from '../../../common/framework/util/util.js';
+import { kTextureFormatInfo } from '../../capability_info.js';
+import { align, isAligned } from '../math.js';
+
+export const kBytesPerRowAlignment = 256;
+export const kBufferCopyAlignment = 4;
+
+export function getTextureCopyLayout(
+  format: GPUTextureFormat,
+  dimension: GPUTextureDimension,
+  size: [number, number, number],
+  mipLevel: number = 0,
+  bytesPerRow?: number,
+  rowsPerImage?: number
+): {
+  bytesPerBlock: number;
+  byteLength: number;
+  minBytesPerRow: number;
+  bytesPerRow: number;
+  rowsPerImage: number;
+} {
+  let mipSize: [number, number, number];
+  switch (dimension) {
+    case '1d':
+      mipSize = [size[0] >> mipLevel, size[1], size[2]];
+      break;
+    case '2d':
+      mipSize = [size[0] >> mipLevel, size[1] >> mipLevel, size[2]];
+      break;
+    case '3d':
+      mipSize = [size[0] >> mipLevel, size[1] >> mipLevel, size[2] >> mipLevel];
+      break;
+    default:
+      unreachable();
+  }
+
+  const { blockWidth, blockHeight, bytesPerBlock } = kTextureFormatInfo[format];
+  assert(!!bytesPerBlock && !!blockWidth && !!blockHeight);
+
+  const minBytesPerRow = (mipSize[0] / blockWidth) * bytesPerBlock;
+  const alignedMinBytesPerRow = align(minBytesPerRow, kBytesPerRowAlignment);
+  if (bytesPerRow !== undefined) {
+    assert(bytesPerRow >= alignedMinBytesPerRow);
+    assert(isAligned(bytesPerRow, kBytesPerRowAlignment));
+  } else {
+    bytesPerRow = alignedMinBytesPerRow;
+  }
+
+  if (rowsPerImage !== undefined) {
+    assert(rowsPerImage >= size[1]);
+  } else {
+    rowsPerImage = size[1];
+  }
+
+  const bytesPerSlice = (bytesPerRow * rowsPerImage) / blockWidth;
```

Comment:
nit: `bytesPerRow * (rowsPerImage / blockHeight)`?
nit: assert rowsPerImage is aligned to blockHeight

Reply:
Hm.. I may be missing some rounding as well. I looked for what we've decided on block-compressed copies and found https://github.com/gpuweb/gpuweb/issues/363 which doesn't look resolved.

like TBD on whether you have a small mip like 15x15, you actually have to round up and copy a 16x16 block